src/c_api/c_api.cc

```diff
@@ -567,23 +569,43 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
 XGB_DLL int XGBoosterLoadModel(BoosterHandle handle, const char* fname) {
   API_BEGIN();
   CHECK_HANDLE();
-  std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname, "r"));
-  static_cast<Learner*>(handle)->Load(fi.get());
+  if (common::FileExtension(fname) == "json") {
+    auto str = common::LoadSequentialFile(fname);
+    CHECK_GT(str.size(), 2);
+    CHECK_EQ(str[0], '{');
+    Json in { Json::Load({str.c_str(), str.size()}) };
+    static_cast<Learner*>(handle)->LoadModel(in);
+  } else {
+    std::unique_ptr<dmlc::Stream> fi(dmlc::Stream::Create(fname, "r"));
+    static_cast<Learner*>(handle)->Load(fi.get());
+  }
   API_END();
 }
 
-XGB_DLL int XGBoosterSaveModel(BoosterHandle handle, const char* fname) {
+XGB_DLL int XGBoosterSaveModel(BoosterHandle handle, const char* c_fname) {
   API_BEGIN();
   CHECK_HANDLE();
-  std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(fname, "w"));
-  auto *bst = static_cast<Learner*>(handle);
-  bst->Save(fo.get());
+  std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(c_fname, "w"));
+  auto *learner = static_cast<Learner *>(handle);
+  learner->Configure();
+  if (common::FileExtension(c_fname) == "json") {
+    Json out { Object() };
+    learner->SaveModel(&out);
+    std::string str;
+    Json::Dump(out, &str);
+    fo->Write(str.c_str(), str.size());
+  } else {
+    auto *bst = static_cast<Learner*>(handle);
+    bst->Save(fo.get());
+  }
   API_END();
 }
 
+// The following two functions are `Load` and `Save` for memory based serialization
+// methods. E.g. Python pickle.
```

Comment:
Note. We may consider using JSON for Python pickling in the future.

Reply:
@hcho3 See documents in : https://github.com/dmlc/xgboost/pull/4732  Specifically the `saving_model.rst` file.

Reply:
@trivialfis So `enable_experimental_json_serialization` lets you use JSON for pickling. Thanks!